extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/VertxGrpcExporter.java

```diff
@@ -0,0 +1,364 @@
+    private static final String GRPC_STATUS = "grpc-status";
+    private static final String GRPC_MESSAGE = "grpc-message";
+
+    private static final Logger internalLogger = Logger.getLogger(VertxGrpcExporter.class.getName());
+
+    private final ThrottlingLogger logger = new ThrottlingLogger(internalLogger); // TODO: is there something in JBoss Logging we can use?
+
+    // We only log unimplemented once since it's a configuration issue that won't be recovered.
+    private final AtomicBoolean loggedUnimplemented = new AtomicBoolean();
+    private final AtomicBoolean isShutdown = new AtomicBoolean();
+    private final String type;
+    private final ExporterMetrics exporterMetrics;
+    private final SocketAddress server;
+    private final boolean compressionEnabled;
+    private final Map<String, String> headers;
+
+    private final GrpcClient client;
+
+    VertxGrpcExporter(
+            String exporterName,
+            String type,
+            Supplier<MeterProvider> meterProviderSupplier,
+            URI grpcBaseUri, boolean compressionEnabled,
+            Duration timeout,
+            Map<String, String> headersMap,
+            Vertx vertx) {
+        this.type = type;
+        this.exporterMetrics = ExporterMetrics.createGrpcOkHttp(exporterName, type, meterProviderSupplier);
+        this.server = SocketAddress.inetSocketAddress(getPort(grpcBaseUri), grpcBaseUri.getHost());
+        this.compressionEnabled = compressionEnabled;
+        this.headers = headersMap;
+        var httpClientOptions = new HttpClientOptions()
+                .setHttp2ClearTextUpgrade(false) // needed otherwise connections get closed immediately
+                .setReadIdleTimeout((int) timeout.getSeconds())
+                .setTracingPolicy(TracingPolicy.IGNORE); // needed to avoid tracing the calls from this gRPC client
+        this.client = GrpcClient.client(vertx, httpClientOptions);
+    }
+
+    private static int getPort(URI uri) {
+        int originalPort = uri.getPort();
+        if (originalPort > -1) {
+            return originalPort;
+        }
+
+        if ("https".equals(uri.getScheme().toLowerCase(Locale.ENGLISH))) {
+            return 443;
+        }
+        return 80;
+    }
+
+    private CompletableResultCode export(TraceRequestMarshaler marshaler, int numItems) {
+        if (isShutdown.get()) {
+            return CompletableResultCode.ofFailure();
+        }
+
+        exporterMetrics.addSeen(numItems);
+
+        var result = new CompletableResultCode();
+        var onSuccessHandler = new ClientRequestOnSuccessHandler(headers, compressionEnabled, exporterMetrics, marshaler,
+                loggedUnimplemented, logger, type, numItems, result);
+        client.request(server)
+                .onSuccess(onSuccessHandler)
+                .onFailure(new Handler<>() {
+                    @Override
+                    public void handle(Throwable t) {
+                        // TODO: is there a better way todo retry?
```

Comment:
retries can be tricky, especially with exponential backoff.

The idea is to capture the action as a Supplier<Future> and then use something like:

```java
var retry = new RetryContainer(3, Duration.ofSecond(1), Duration.ofSecond(10));
retry.retry(supplier);)
```

retry would invoke the supplier check for the outcome and then use a vertx timer to schedule the next attempt.

Reply:
I did not know this class.

I'll look into using it tomorrow

Reply:
Ah, I thought it existed in Vert.x.

Do we have that somewhere else?

Reply:
Is there perhaps a Mutiny version of the Vert.x gRPC Client that I could use and make things nicer?

Reply:
I'm actually not sure... the old one I think so, but maybe not the new one. @jponge do you know? I don't remember adding it.

Reply:
We don't have that

Reply:
No problem. In that case, I'll leave the PR as is for now and we can certainly improve in the future